Feature/recruitments list

daeun0608 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #32
- 작업 요약: 스터디 구인 공고 목록 조회에 사용되는 시리얼라이저 작성

## 📄 상세 내용
- [x] 스터디 구인 공고 목록 조회에 사용되는 시리얼라이저 작성

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

daeun0608 committed:
✨  feat: 스터디 구인 공고 목록 조회에 사용되는 시리얼라이저 작성
스터디 구인 공고 목록 조회에 사용되는 시리얼라이저 작성
관련 이슈: #32

daeun0608 committed:
🎨 style: black, isort 코드 포매터 실행
black, isort 코드 포매터 실행
관련 이슈: #32

SimJunSik commented on apps/recruitments/serializers/serializers_list.py:
```diff
@@ -0,0 +1,51 @@
+    def get_lectures(self, obj):
+        result = []
+
+        for study_group in obj.study_group.all():
+            lectures = study_group.lectures.all()
```
N+1 query 에 대해 미리 고려가 되면 좋을 것 같습니다 ( 되어있다면 👍 )

daeun0608 replied:
안녕하세요 제가 전에 쿼리 최적화를 경험해 본적이 있는데 그때는 뷰에서 prefetch_related를 사용해서 해당 N+1문제를 해결했었는데 혹시 같은 방법으로 문제를 해결해도 될까요?

daeun0608 replied:
쿼리 최적화를 했을때 해당 함수에 문제가 있는거 같긴해서 후에 pr올리기 전까지 함수는 좀 더 개선해보겠습니다.

SimJunSik replied:
네 그 방법으로 하셔도 되고, 이것저것 시도해보시고 본인만의 결론을 내보려고 하는게 제일 좋습니다
아시겠지만, 면접이나 실무에서도 여러 방법이 있을 때 각각의 장단점 혹은 현재 상황에 알맞는 방법이 어떤 건지 의견을 내거나 생각해볼 줄 아는게 중요해서 쪼끔 시간이 걸리더라도 한 번 시도해보세요!

SimJunSik commented on apps/recruitments/serializers/serializers_list.py:
```diff
@@ -0,0 +1,51 @@
+        )
+
+    def get_lectures(self, obj):
+        result = []
```
study_group 마다 중복되는 lecture 를 가지는 경우가 없을까요? 그런 경우가 있다면, list 말고 다른 자료구조(자료형) 를 고려해보시는 게 좋을 것 같네요

daeun0608 replied:
study_group과 lecture가 매핑되어있는 테이블이라서 중복되지는 않을 것 같습니다!

daeun0608 commented:
mypy문제로 잠시 close 했습니다. 내일 수정해서 다시 Pr하겠습니다.

daeun0608 committed:
🎨 style: mypy 통과
mypy 통과
관련 이슈: #32

daeun0608 committed:
🎨 style: black, isort 코드 포매터 실행
black, isort 코드 포매터 실행
관련 이슈: #32

Meoyoug commented on apps/recruitments/serializers/serializers_list.py:
```diff
@@ -0,0 +1,49 @@
+
+
+# 비직관적인 이름 변경과 다대다 관계 처리
+class LectureSerializer(serializers.ModelSerializer[Lecture]):
```
조금더 디테일하게!
LectureListSerializer

Umdoong reviewed (dismissed):
고생하셨습니다.

Meoyoug commented on apps/recruitments/serializers/serializers_list.py:
```diff
@@ -0,0 +1,49 @@
+    # recruiments-study_groups-study_lectures-crawled_lectures > study_group__lectures로 최적화
+    # queryset.select_related('study_group').prefetch_related('study_group__lectures')
+    # 쿼리 최적화에 사용( https://docs.djangoproject.com/en/5.2/ref/models/querysets/#only )
+    lectures = serializers.SerializerMethodField(read_only=True)
```
SerializerMethodField는 유용하지만 Swagger-ui에서 스키마에 반영되지 않아요
NestedSerializer를 활용해보는게 좋을 것 같아요!

<img width="1163" height="966" alt="image" src="https://github.com/user-attachments/assets/c79c272b-8e54-4ae6-b251-ab219fd2ed32" />
https://www.django-rest-framework.org/api-guide/relations/

Meoyoug commented on apps/recruitments/serializers/serializers_list.py:
```diff
@@ -0,0 +1,49 @@
+    class Meta:
+        model = Recruitment
+        fields = (
+            "id",
```
uuid도 포함해주세요~!

Meoyoug commented on apps/recruitments/serializers/serializers_list.py:
```diff
@@ -0,0 +1,49 @@
+        fields = ("title", "instructor")
+
+
+class RecruitmentSerializer(serializers.ModelSerializer[Recruitment]):
```
이것도 아마 RecruitmentListSerializer가 명확할 것 같아요!

Meoyoug commented on apps/recruitments/serializers/serializers_list.py:
```diff
@@ -0,0 +1,49 @@
+    # bookmarks_count는 bookmark_users의 개수만 받기 ( 조회 때 추가 )
+    bookmarks_count = serializers.IntegerField(read_only=True)
+
+    class Meta:
```
List시리얼라이저면 읽기 전용이 될테니 모든 필드에 대해서 read_only 설정을 추가해보기

daeun0608 committed:
♻️refactor : 시리얼라이저 이름 변경 및 전체 read_only설정
- class이름에 List를 추가해 디테일하게 표기
- Swagger-ui에 사용못하는 SerializerMethodField변경
- uuid 추가
- 조회기능이기 때문에 read_only설정을 전체적으로 적용
관련 이슈: #32

Meoyoug reviewed (approved):
고생하셨습니다 머지할게요~
